### ISSUE TITLE
Running on iOS

### ISSUE THREAD

**Issue by RyosukeSakaguchi** [ultralytics/yolov5#892]
## ❔Question
I'm trying to convert the YOLOv5 model into coreML and run it on iOS.
However, coreML does not include the processes in detect.py.
https://github.com/ultralytics/yolov5/blob/master/detect.py
In other words, I need to process coreML's output as done in detect.py.
Therefore, I would like to have a Swift version of detect.py processing code (including NMS etc.) to run on iOS.
Can I have it?

**Comment by spicybeef003** [comment]
i would be interested in this too if possible!

**Comment by jenniferjiangkells** [comment]
I would also like to see a tutorial on the pipeline for exporting to iOS!

**Comment by AlvinZheng** [comment]
me too

**Comment by radu-diaconescu13** [comment]
I am also interested in this

**Comment by MaddyThakker** [comment]
any update on this?

**Comment by bradgrimm** [comment]
I ran into this same problem, and as a result I put together a working prototype:
https://github.com/bradgrimm/YOLO-CoreML.

It requires an export such as:
```
python models/export.py --train --weights runs/train/exp20/weights/last.pt --img 640 --batch 1
```

And then a handful of values need to be updated such as the inputWidth/inputHeight, anchors, and strides variables. There's python code in the comments of the swift on how to grab those from your model. Hopefully this helps anyone who is struggling trying to convert from the yolo features to bounding boxes in an iOS app.

This is forked from a previous version by [syshen](https://github.com/syshen/YOLO-CoreML)

**Comment by NaeemKhan333** [comment]
@glenn-jocher Any update on this?

**Comment by iscyy** [comment]
> I ran into this same problem, and as a result I put together a working prototype: https://github.com/bradgrimm/YOLO-CoreML.
> 
> It requires an export such as:
> 
> ```
> python models/export.py --train --weights runs/train/exp20/weights/last.pt --img 640 --batch 1
> ```
> 
> And then a handful of values need to be updated such as the inputWidth/inputHeight, anchors, and strides variables. There's python code in the comments of the swift on how to grab those from your model. Hopefully this helps anyone who is struggling trying to convert from the yolo features to bounding boxes in an iOS app.
> 
> This is forked from a previous version by [syshen](https://github.com/syshen/YOLO-CoreML)

Hi, how should the inputWidth/inputHeight, anchors and strides variables be set? Is there a complete solution?

**Comment by bradgrimm** [comment]
@Him-wen A full example in Python would look something like:
```
import torch
from models.yolo import Detect

model = torch.load('models/2021_06_21.pt')
model = model['model']

for module in model.modules():
    if isinstance(module, Detect):
        detect = module
        break

print(f'Stride: {detect.stride.tolist()}')
print(f'Anchors: {detect.anchor_grid.squeeze().tolist()}')

grid_size = detect.grid[0].shape[2:4]
stride = detect.stride[0].numpy()
print(f'Input size: {grid_size * stride}')
```

Then from there you can just copy those values into [Yolo.swift](https://github.com/bradgrimm/YOLO-CoreML/blob/master/YOLO-CoreML/YOLO-CoreML/YOLO.swift)

**Comment by iscyy** [comment]
> 



> @Him-wen A full example in Python would look something like:
> 
> ```
> import torch
> from models.yolo import Detect
> 
> model = torch.load('models/2021_06_21.pt')
> model = model['model']
> 
> for module in model.modules():
>     if isinstance(module, Detect):
>         detect = module
>         break
> 
> print(f'Stride: {detect.stride.tolist()}')
> print(f'Anchors: {detect.anchor_grid.squeeze().tolist()}')
> 
> grid_size = detect.grid[0].shape[2:4]
> stride = detect.stride[0].numpy()
> print(f'Input size: {grid_size * stride}')
> ```
> 
> Then from there you can just copy those values into [Yolo.swift](https://github.com/bradgrimm/YOLO-CoreML/blob/master/YOLO-CoreML/YOLO-CoreML/YOLO.swift)

Hello, I tried according to your suggestion, but still did not get the correct result and reported an error, can you provide the complete [code](https://github.com/bradgrimm/YOLO-CoreML) that can run normally in your configuration environment and model.pt and get the result? Run first, then study, thank you for your advice

**Comment by bradgrimm** [comment]
@Him-wen I uploaded model that works with that repository [here](https://github.com/bradgrimm/YOLO-CoreML/blob/master/Models/2021_06_21.pt).

But please note that my fork was modified to detect lego pieces. So that model likely doesn't do what you want. And there are some changes to that repository related to that change.

Labels names were changed to lego labels in [Helpers.swift](https://github.com/bradgrimm/YOLO-CoreML/blob/master/YOLO-CoreML/YOLO-CoreML/Helpers/Helpers.swift)

Layers are named differently than the ones used in the syshen model (now they are layers _714, _727, _740) in [Yolo.swift](https://github.com/bradgrimm/YOLO-CoreML/blob/master/YOLO-CoreML/YOLO-CoreML/YOLO.swift)

So it may take a little elbow grease to get that repository working for a non-lego detection model.
